tests/test_init_lazy_loading.py

```diff
@@ -0,0 +1,151 @@
+        )
+        reordered_content_before_static_checks = SUBMOD_ATTRS_PATTERN.sub(
+            _submod_attrs_definition, init_content_before_static_checks
+        )
+
+        # Generate the static imports given the `_SUBMOD_ATTRS` dictionary.
+        static_imports = [
+            f"    from .{module} import {attr} # noqa: F401"
+            for module, attributes in _SUBMOD_ATTRS.items()
+            for attr in attributes
+        ]
+
+        # Generate the expected `__init__.py` file content and apply formatter on it.
+        expected_init_content = isort.code(
+            reordered_content_before_static_checks
+            + IF_TYPE_CHECKING_LINE
+            + "\n".join(static_imports)
+            + "\n",
+            config=isort.Config(settings_path=SETUP_CFG_PATH),
+        )
+
+        # If expected `__init__.py` content is different, test fails. If '--update-init-file'
+        # is used, `__init__.py` file is updated before the test fails.
+        if init_content != expected_init_content:
+            if self.update_init_file:
+                with init_path.open("w") as f:
+                    f.write(expected_init_content)
+
+                self.fail(
+                    "Pytest was run with '--update-init-file' option and"
+                    " `./src/huggingface_hub/__init__.py` content has been updated. It"
+                    " is most likely that you added a module/function to"
+                    " `_SUBMOD_ATTRS` and did not update the 'static import'-part."
+                    " Please make sure the changes are accurate and if yes, commit them"
+                    " and re-run this test without the '--update-init-file' option."
+                )
+            else:
+                self.fail(
+                    "Expected content mismatch in `./src/huggingface_hub/__init__.py`."
+                    " It is most likely that you added a module/function to"
+                    " `_SUBMOD_ATTRS` and did not update the 'static import'-part. To"
+                    " do it, please re-run the test suite with '--update-init-file'"
+                    " option: `pytest tests/test_init_lazy_loading.py -k"
+                    " test_static_imports --update-init-file`. Look at the changes and"
+                    " if accurate, commit them and re-run this test without the"
+                    " '--update-init-file' option."
+                )
+
+        # Should never fail but let's check it just in case
+        self.assertEqual(init_content, expected_init_content)
+
+    def test_autocomplete_on_root_imports(self) -> None:
+        """Test autocomplete with `huggingface_hub` works with Jedi.
+
+        Not all autocomplete systems are based on Jedi but if this one works we can
+        assume others does it as well.
+        """
+        source = """from huggingface_hub import c"""
+        script = jedi.Script(source, path="example.py")
+        completions = script.complete(1, len(source))
```

Comment:
slightly off topic, and I'm really far from being an expert, but isn't pylance or pyright taking more steam than Jedi now? (not sure what the main ≠ are)

Reply:
in all cases pretty cool that you're able to test cover autocomplete 🤯

Reply:
Really not an expert either on that :grin: 
Jedi was easier to integrate in the test suite that's all :)